src/metabase/api/database.clj

```diff
@@ -261,22 +270,116 @@
+      (do
+        (log/error ex "Error while streaming results for response")
+        (throw ex))
+      ;; The error occured before we have received any results from the query, deliver the exception so that an error
+      ;; response can be returned
+      (deliver p ex))))
+
+(defn- rename-fields-for-response
+  "Munges keypairs coming from the database into what the UI expects in the resposne"
+  [{:keys [id display_name table base_type special_type]}]
+  {:id           id
+   :name         display_name
+   :base_type    base_type
+   :special_type special_type
+   :table_name   (:display_name table)
+   :schema       (:schema table)})
+
+(def ^:private fields-xform
+  (comp (filter mi/can-read?)
+        ;; This will create 10,000 row chunks of data, being eager for that 10,000 row chunk. This will allow the
+        ;; hydrate calls to be streamed along with the regular database results. Not sure what the best value is for
+        ;; this. A lower value is slower but more memory efficient, a higher value is faster but will consume more
+        ;; memory
+        (partition-all 10000)
+        ;; Hydrate in 10,000 row batches
+        (map #(hydrate % :table))
+        ;; This will flatten the chunks about out
+        cat
+        ;; Munge the keypair names to what the UI expects
+        (map rename-fields-for-response)))
+
 (api/defendpoint GET "/:id/fields"
   "Get a list of all `Fields` in `Database`."
   [id]
   (api/read-check Database id)
-  (let [fields (filter mi/can-read? (-> (db/select [Field :id :display_name :table_id :base_type :special_type]
-                                          :table_id        [:in (db/select-field :id Table, :db_id id)]
-                                          :visibility_type [:not-in ["sensitive" "retired"]])
-                                        (hydrate :table)))]
-    (for [{:keys [id display_name table base_type special_type]} fields]
-      {:id           id
-       :name         display_name
-       :base_type    base_type
-       :special_type special_type
-       :table_name   (:display_name table)
-       :schema       (:schema table)})))
-
+  (let [p        (promise)
+        response (serialize-to-piped-input-stream #(deliver p %) (make-error-handler p) fields-xform
+                                                  (db/select-reducible [Field :id :display_name :table_id :base_type :special_type]
+                                                    :table_id        [:in (db/select-field :id Table, :db_id id)]
+                                                    :visibility_type [:not-in ["sensitive" "retired"]]))]
+
+    ;; Block until a row or an exception has been delivered to `p`. This will catch something like SQL query that
+    ;; fails to compile or references something that doesn't exist and allow that exception to be returned as if the
+    ;; query was running in the HTTP thread.
+    (throw-if-exception p)
```

Comment:
This is going to be super subjective: but this approach with promises and cross-thread communication is making my brain hurt. Couldn't this be done with something state-machine like?